File: tests/accounting/readiness.php
<?php
require __DIR__.'/../../yiimp2/services/ZclNodeReadiness.php';
use app\services\ZclNodeReadiness;
function check(bool $condition,string $label): void {if (!$condition) throw new RuntimeException($label);}
$now=1800000000;
$chain=['chain'=>'main','blocks'=>3200000,'headers'=>3200000,'bestblockhash'=>str_repeat('a',64),'verificationprogress'=>1,
    'bootstrap_validation'=>['tip_hold'=>false],'finalization_hold'=>['held'=>false]];
$header=['hash'=>str_repeat('a',64),'height'=>3200000,'time'=>$now-75,'confirmations'=>1];
check(ZclNodeReadiness::ready($chain,$header,3,$now),'current corroborated node is ready');
foreach ([['chain','regtest'],['headers',3199999],['blocks','3200000'],['verificationprogress',0.99],['initialblockdownload',true],['bootstrap_validation',[]],['bootstrap_validation',['tip_hold'=>true]],['finalization_hold',['held'=>true]]] as [$field,$value]) {
    $bad=$chain;$bad[$field]=$value;check(!ZclNodeReadiness::ready($bad,$header,3,$now),'reject '.$field);
}
foreach ([['hash',str_repeat('b',64)],['height',3199999],['time',$now-1801],['time',$now+601],['confirmations',-1]] as [$field,$value]) {
    $bad=$header;$bad[$field]=$value;check(!ZclNodeReadiness::ready($chain,$bad,3,$now),'reject header '.$field);
}
check(!ZclNodeReadiness::ready($chain,$header,0,$now),'reject disconnected node');
echo "ZCL mainnet readiness guards passed.\n";

File: tests/public-data/readiness.py
import copy
import importlib.util
from pathlib import Path
import unittest

path = Path(__file__).resolve().parents[2] / 'deploy/zcl/publish-richlist.py'
spec = importlib.util.spec_from_file_location('richlist_publisher', path)
publisher = importlib.util.module_from_spec(spec)
spec.loader.exec_module(publisher)

class Readiness(unittest.TestCase):
    def fixture(self):
        return {'chain': 'main', 'blocks': 3247730, 'headers': 3247730, 'bestblockhash': 'a'*64,
            'verificationprogress': 1, 'bootstrap_validation': {'state': 'disabled', 'tip_hold': False},
            'finalization_hold': {'held': False}}, {'height': 3247730, 'hash': 'a'*64, 'time': 1789190000}

    def test_current_anchor_and_genesis_validation(self):
        chain, header = self.fixture()
        publisher.require_current_chain(chain, header, header['time'] + 60)
        chain['bootstrap_validation']['state'] = 'validated'
        publisher.require_current_chain(chain, header, header['time'] + 60)

    def test_sync_and_active_or_unknown_validation_holds(self):
        mutations = [lambda c: c.update(blocks=3247729), lambda c: c.update(chain='regtest'),
            lambda c: c.update(verificationprogress=float('nan')), lambda c: c.update(verificationprogress=.99),
            lambda c: c.update(verificationprogress=1.01), lambda c: c.update(initialblockdownload=True),
            lambda c: c['bootstrap_validation'].update(tip_hold=True),
            lambda c: c['bootstrap_validation'].update(state='provisional'),
            lambda c: c.update(bootstrap_validation={}), lambda c: c['finalization_hold'].update(held=True),
            lambda c: c.update(finalization_hold={})]
        for change in mutations:
            chain, header = self.fixture()
            change(chain)
            with self.assertRaises(RuntimeError):
                publisher.require_current_chain(chain, header, header['time'] + 60)

    def test_stale_future_and_mismatched_header(self):
        chain, header = self.fixture()
        for now in [header['time'] + 1801, header['time'] - 301]:
            with self.assertRaises(RuntimeError): publisher.require_current_chain(chain, header, now)
        wrong = copy.copy(header); wrong['hash'] = 'b'*64
        with self.assertRaises(RuntimeError): publisher.require_current_chain(chain, wrong, header['time'])

if __name__ == '__main__': unittest.main()
